apps/ledger-live-mobile/src/components/Web3AppWebview/helpers.ts

```diff
@@ -0,0 +1,178 @@
+};
+
+type useWebviewStateParams = {
+  manifest: AppManifest;
+  inputs?: Record<string, string>;
+};
+
+export function useWebviewState(
+  params: useWebviewStateParams,
+  WebviewAPIRef: React.ForwardedRef<WebviewAPI>,
+) {
+  const webviewRef = useRef<WebView>(null);
+  const { manifest, inputs } = params;
+
+  const initialURL = useMemo(() => {
+    const url = new URL(manifest.url);
+    addParamsToURL(url, inputs);
+    if (manifest.params) {
+      url.searchParams.set("params", JSON.stringify(manifest.params));
+    }
+    return url.toString();
+  }, [manifest, inputs]);
+
+  const [state, setState] = useState<WebviewState>(initialWebviewState);
+
+  const [currentURI, setURI] = useState(initialURL);
+  const { theme } = useTheme();
+
+  const source = useMemo(
+    () => ({
+      uri: currentURI,
+      headers: getClientHeaders({
+        client: "ledger-live-mobile",
+        theme,
+      }),
+    }),
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [currentURI],
+  );
+
+  useImperativeHandle(
+    WebviewAPIRef,
+    () => {
+      return {
+        reload: () => {
+          const webview = safeGetRefValue(webviewRef);
+
+          webview.reload();
+        },
+        goBack: () => {
+          const webview = safeGetRefValue(webviewRef);
+
+          webview.goBack();
+        },
+        goForward: () => {
+          const webview = safeGetRefValue(webviewRef);
+
+          webview.goForward();
+        },
+        loadURL: (url: string): void => {
+          setURI(url);
+        },
+      };
+    },
+    [webviewRef],
+  );
+
+  const onLoad: Required<WebViewProps>["onLoad"] = useCallback(
+    ({ nativeEvent }) => {
+      setState({
+        title: nativeEvent.title,
+        url: nativeEvent.url,
+        canGoBack: nativeEvent.canGoBack,
+        canGoForward: nativeEvent.canGoForward,
+        loading: nativeEvent.loading,
+      });
+    },
+    [],
+  );
+
+  const onLoadStart: Required<WebViewProps>["onLoadStart"] = useCallback(
+    ({ nativeEvent }) => {
+      setState({
+        title: nativeEvent.title,
+        url: nativeEvent.url,
+        canGoBack: nativeEvent.canGoBack,
+        canGoForward: nativeEvent.canGoForward,
+        loading: nativeEvent.loading,
+      });
+    },
+    [],
+  );
+
+  const onLoadEnd: Required<WebViewProps>["onLoadEnd"] = useCallback(
+    ({ nativeEvent }) => {
+      setState({
+        title: nativeEvent.title,
+        url: nativeEvent.url,
+        canGoBack: nativeEvent.canGoBack,
+        canGoForward: nativeEvent.canGoForward,
+        loading: nativeEvent.loading,
+      });
+    },
+    [],
+  );
+
+  const onLoadProgress: Required<WebViewProps>["onLoadProgress"] = useCallback(
+    ({ nativeEvent }) => {
+      setState({
+        title: nativeEvent.title,
+        url: nativeEvent.url,
+        canGoBack: nativeEvent.canGoBack,
+        canGoForward: nativeEvent.canGoForward,
+        loading: nativeEvent.loading,
+      });
+    },
+    [],
+  );
```

Comment:
We can probably define only 1 function and reuse it for every `useCallback` because they are all the same

Reply:
True, and I thought about it.

But then I arbitrarily had to use one of the types for every single function and it felt wrong 😂 
Feels more readable like that since there aren't so many :)